mmengine/visualization/vis_backend.py

```diff
@@ -610,3 +611,176 @@ def close(self):
+        >>> vis_backend.add_scaler('mAP', 0.6)
+        >>> vis_backend.add_scalars({'loss': 0.1,'acc':0.8})
+        >>> cfg = Config(dict(a=1, b=dict(b1=[0, 1])))
+        >>> vis_backend.add_config(cfg)
+
+    Args:
+        save_dir (str): The root directory to save the files
+            produced by the backend.
+        exp_name (str, optional): The experiment name. Default to None.
+        run_name (str, optional): The run name. Default to None.
+        tags (dict, optional): The tags to be added to the experiment.
+            Default to None.
+        params (dict, optional): The params to be added to the experiment.
+            Default to None.
+        tracking_uri (str, optional): The tracking uri. Default to None.
+        log_artifact (bool, optional): Whether to log the artifact.
+            Default to False.
+    """
+
+    def __init__(self,
+                 save_dir: str,
+                 exp_name: Optional[str] = None,
+                 run_name: Optional[str] = None,
+                 tags: Optional[dict] = None,
+                 params: Optional[dict] = None,
+                 tracking_uri: Optional[str] = None,
+                 log_artifact: bool = False):
+        super().__init__(save_dir)
+        self._exp_name = exp_name
+        self._run_name = run_name
+        self._tags = tags
+        self._params = params
+        self._tracking_uri = tracking_uri
+        self._log_artifact = log_artifact
+
+    def _init_env(self):
+        """Setup env for MLFlow."""
+        if not os.path.exists(self._save_dir):
+            os.makedirs(self._save_dir, exist_ok=True)  # type: ignore
+
+        try:
+            import mlflow
+        except ImportError:
+            raise ImportError(
+                'Please run "pip install mlflow" to install mlflow'
+            )  # type: ignore
+        self._mlflow = mlflow
+
+        if self._tracking_uri is not None:
+            self._mlflow.set_tracking_uri(self._tracking_uri)
+        else:
+            self._mlflow.set_tracking_uri(f'file://{self._save_dir}')
+
+        if self._mlflow.get_experiment_by_name(self._exp_name) is None:
+            self._mlflow.create_experiment(self._exp_name)
+
+        self._mlflow.set_experiment(self._exp_name)
+
+        if self._run_name is not None:
+            self._mlflow.set_tag('mlflow.runName', self._run_name)
+        if self._tags is not None:
+            self._mlflow.set_tags(self._tags)
+        if self._params is not None:
+            self._mlflow.log_params(self._params)
+
+    @property  # type: ignore
+    @force_init_env
+    def experiment(self):
+        """Return MLFlow object."""
+        return self._mlflow
+
+    @force_init_env
+    def add_config(self, config: Config, **kwargs) -> None:
+        """Record the config to mlflow.
+
+        Args:
+            config (Config): The Config object
+        """
+        self.cfg = config
+        self._mlflow.log_params(self._flatten(self.cfg))
```

Comment:
Should we directly use `log_text` which is consistent with `Tensorboard`? https://mlflow.org/docs/latest/python_api/mlflow.html#mlflow.log_text

Reply:
Not sure if I should open an issue or comment here directly:
With this branch, logging on a mlflow server (version 2.1.1) causes the following error: `mlflow.exceptions.RestException: INVALID_PARAMETER_VALUE: Param value '[{'type': 'LoadImageFromFile', 'file_client_args': {'backend': 'disk'}}, {'type': 'LoadAnnotations', 'with_bbox': True}, {'type': 'CachedMosaic', 'img_scale': (640, 640), 'pad_val': 114.0}, {'type': 'RandomResize', 'scale': (1280, 1280), 'ratio_range' had length 633, which exceeded length limit of 500`
This limit was added from the mlflow version 1.28 and on...
config used:
`_base_ = "./rtmdet-ins_l_8xb32-300e_coco.py"

model = dict(
    backbone=dict(deepen_factor=0.67, widen_factor=0.75),
    neck=dict(in_channels=[192, 384, 768], out_channels=192, num_csp_blocks=2),
    bbox_head=dict(in_channels=192, feat_channels=192),
)


train_pipeline = [
    dict(type="LoadImageFromFile", file_client_args={{_base_.file_client_args}}),
    dict(type="LoadAnnotations", with_bbox=True),
    dict(type="CachedMosaic", img_scale=(640, 640), pad_val=114.0),
    dict(
        type="RandomResize", scale=(1280, 1280), ratio_range=(0.1, 2.0), keep_ratio=True
    ),
    dict(type="RandomCrop", crop_size=(640, 640)),
    dict(type="YOLOXHSVRandomAug"),
    dict(type="RandomFlip", prob=0.5),
    dict(type="Pad", size=(640, 640), pad_val=dict(img=(114, 114, 114))),
    dict(
        type="CachedMixUp",
        img_scale=(640, 640),
        ratio_range=(1.0, 1.0),
        max_cached_images=20,
        pad_val=(114, 114, 114),
    ),
    dict(type="PackDetInputs"),
]

train_pipeline_stage2 = [
    dict(type="LoadImageFromFile", file_client_args={{_base_.file_client_args}}),
    dict(type="LoadAnnotations", with_bbox=True),
    dict(
        type="RandomResize", scale=(640, 640), ratio_range=(0.1, 2.0), keep_ratio=True
    ),
    dict(type="RandomCrop", crop_size=(640, 640)),
    dict(type="YOLOXHSVRandomAug"),
    dict(type="RandomFlip", prob=0.5),
    dict(type="Pad", size=(640, 640), pad_val=dict(img=(114, 114, 114))),
    dict(type="PackDetInputs"),
]


train_dataloader = dict(
    batch_size=16,
    num_workers=4,
    batch_sampler=None,
    pin_memory=True,
    dataset=dict(pipeline=train_pipeline),
)

max_epochs = 150
stage2_num_epochs = 20
base_lr = 0.004
interval = 10

train_cfg = dict(
    max_epochs=max_epochs,
    val_interval=interval,
    dynamic_intervals=[(max_epochs - stage2_num_epochs, 1)],
)


# learning rate
param_scheduler = [
    dict(type="LinearLR", start_factor=1.0e-5, by_epoch=False, begin=0, end=1000),
    dict(
        # use cosine lr from 150 to 300 epoch
        type="CosineAnnealingLR",
        eta_min=base_lr * 0.05,
        begin=max_epochs // 2,
        end=max_epochs,
        T_max=max_epochs // 2,
        by_epoch=True,
        convert_to_iter_based=True,
    ),
]
visualizer = dict(
    type="DetLocalVisualizer",
    vis_backends=[
        dict(type="LocalVisBackend"),
        dict(
            type="MLFlowVisBackend",
            # run_name="xxx",
            exp_name="xxxxx",
            tags=dict(name="xxxx", type="rtmdet_m"),
            params=dict(dev_param=0),
            tracking_uri="http://x.x.x.x:5000",
        ),
    ],
)
`

Reply:
> Should we directly use which is consistent with ? https://mlflow.org/docs/latest/python_api/mlflow.html#mlflow.log_text`log_text``Tensorboard`

`log_text` is used to store text files in mlflow artifacts. Doesn't seem appropriate.
There is already a source of config in OpenMMLab's work_dir, so it is saved as artifacts without specifying it separately.

![image](https://user-images.githubusercontent.com/66167254/215263474-d9d850b7-3d4a-45cd-b060-3da7ceb694c7.png)

Reply:
@zhouzaida Like Tensorboard, config.py has been modified to save by default.

Reply:
@mchaniotakis Modified to handle `dict` inside `list` as well. Can you check again?

Reply:
No errors now with list objects:
<img width="690" alt="image" src="https://user-images.githubusercontent.com/47221563/215265578-58dc7b15-636b-432c-8c2f-f0959262cb1c.png">